File: event_hub_spring/src/main/java/com/eventhub/entity/Ticket.java
package com.eventhub.entity;

import jakarta.persistence.*;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "tickets")
public class Ticket {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "ticket_id", updatable = false, nullable = false)
    private String ticketId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "event_id")
    private Event event;

    @Column(nullable = false)
    private String type = "standard";

    @Column(nullable = false)
    private Double prix;

    @Column(name = "quantite_disponible", nullable = false)
    private Integer quantiteDisponible;

    @Column(name = "quantite_vendue", nullable = false)
    private Integer quantiteVendue = 0;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false)
    private Boolean actif = true;

    @OneToMany(mappedBy = "ticket", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<Booking> bookings = new ArrayList<>();

    public Ticket() {}

    public String getTicketId() { return ticketId; }
    public void setTicketId(String ticketId) { this.ticketId = ticketId; }
    public Event getEvent() { return event; }
    public void setEvent(Event event) { this.event = event; }
    public String getType() { return type; }
    public void setType(String type) { this.type = type; }
    public Double getPrix() { return prix; }
    public void setPrix(Double prix) { this.prix = prix; }
    public Integer getQuantiteDisponible() { return quantiteDisponible; }
    public void setQuantiteDisponible(Integer quantiteDisponible) { this.quantiteDisponible = quantiteDisponible; }
    public Integer getQuantiteVendue() { return quantiteVendue; }
    public void setQuantiteVendue(Integer quantiteVendue) { this.quantiteVendue = quantiteVendue; }
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
    public Boolean getActif() { return actif; }
    public void setActif(Boolean actif) { this.actif = actif; }
    public List<Booking> getBookings() { return bookings; }
    public void setBookings(List<Booking> bookings) { this.bookings = bookings; }

    public boolean isAvailable() { return actif != null && actif && quantiteDisponible != null && quantiteDisponible > 0; }
    public boolean isSoldOut() { return quantiteDisponible != null && quantiteDisponible <= 0; }
    public boolean isStandard() { return "standard".equals(type); }
    public boolean isVip() { return "vip".equals(type); }
    public boolean isEarlyBird() { return "early_bird".equals(type); }
    public String getTypeDisplay() {
        return switch (type) {
            case "vip" -> "VIP";
            case "early_bird" -> "Early Bird";
            default -> "Standard";
        };
    }
}

File: event_hub_back/lib/models/ticket.dart
class Ticket {
  final String ticketId;
  final String? eventId;
  final String type;
  final double prix;
  final int quantiteDisponible;
  final int quantiteVendue;
  final String description;
  final bool actif;

  Ticket({
    required this.ticketId,
    this.eventId,
    required this.type,
    required this.prix,
    required this.quantiteDisponible,
    this.quantiteVendue = 0,
    this.description = '',
    this.actif = true,
  });

  bool get isAvailable => actif && quantiteDisponible > 0;
  bool get isSoldOut => quantiteDisponible <= 0;
  bool get isStandard => type == 'standard';
  bool get isVip => type == 'vip';
  bool get isEarlyBird => type == 'early_bird';

  String get typeDisplay {
    switch (type) {
      case 'vip':
        return 'VIP';
      case 'early_bird':
        return 'Early Bird';
      default:
        return 'Standard';
    }
  }

  factory Ticket.fromMap(Map<String, dynamic> map) {
    return Ticket(
      ticketId: map['ticketId'] ?? map['id'] ?? '',
      eventId: map['eventId'],
      type: map['type'] ?? 'standard',
      prix: (map['prix'] ?? 0).toDouble(),
      quantiteDisponible: map['quantiteDisponible'] ?? 0,
      quantiteVendue: map['quantiteVendue'] ?? 0,
      description: map['description'] ?? '',
      actif: map['actif'] ?? true,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'ticketId': ticketId,
      'eventId': eventId,
      'type': type,
      'prix': prix,
      'quantiteDisponible': quantiteDisponible,
      'quantiteVendue': quantiteVendue,
      'description': description,
      'actif': actif,
    };
  }
}
